### PR TITLE
Spline #276 Switch to async arango driver

### PR DESCRIPTION
**Pull request by leborgneromain**
It also includes the retries for the rest calls

### CODE REVIEW

File: persistence/src/main/scala/za/co/absa/spline/persistence/Persister.scala [hunk 1]
```diff
@@ -16,110 +16,45 @@
...
-      .map(vpack.deserialize[DataSource](_, classOf[DataSource]))
-      .map(ds => ds.uri -> ds._key.get)
-      .toMap
-  }
-
-  private def generateNewKeys(uris: Seq[String], uriToExistingKey: Map[String, String]) = {
-    (uris.toSet -- uriToExistingKey.keys)
-        .map(_ -> randomUUID.toString)
-        .toMap
+    Try(attemptSave(entity))
```
Comment (wajda) [inline persistence/src/main/scala/za/co/absa/spline/persistence/Persister.scala]: Are we catching errors from the `attemptSave` method only or the code that is executing asynchronously and is referred by the returned future?
What will happen if the future itself is resolved with an error?

File: integration-tests/src/main/scala/za/co/absa/spline/test/fixture/ArangoFixture.scala [hunk 2]
```diff
@@ -16,25 +16,26 @@
...
+import scala.concurrent.Future
 
 trait ArangoFixture extends BeforeAndAfterEach {
 
+
   this: Suite =>
 
+  import za.co.absa.spline.common.ConfigurationImplicits._
+
   val arangoUri: String = new SystemConfiguration().getRequiredString("test.spline.arangodb.url")
```
[the remark this answers is not in the document]
Comment (leborgneromain) [reply integration-tests/src/main/scala/za/co/absa/spline/test/fixture/ArangoFixture.scala]: fixed

File: persistence/src/main/scala/za/co/absa/spline/persistence/ArangoInit.scala [hunk 3]
```diff
@@ -16,38 +16,44 @@
...
-      new EdgeDefinition().collection("progressOf").from("progress").to("execution"),
-      new EdgeDefinition().collection("executes").from("execution").to("operation"),
-      new EdgeDefinition().collection("follows").from("operation").to("operation"),
-      new EdgeDefinition().collection("readsFrom").from("operation").to("dataSource"),
-      new EdgeDefinition().collection("writesTo").from("operation").to("dataSource")
-    ).asJava
-    db.createGraph("lineage", edgeDefs)
+  import scala.concurrent.ExecutionContext.Implicits._
+
+  def initialize(db: ArangoDatabaseAsync, dropIfExists: Boolean) = {
```
[the remark this answers is not in the document]
Comment (leborgneromain) [reply persistence/src/main/scala/za/co/absa/spline/persistence/ArangoInit.scala]: fixed

File: persistence/src/main/scala/za/co/absa/spline/persistence/ArangoInit.scala [hunk 4]
```diff
@@ -16,38 +16,44 @@
...
+    db.exists().toScala.flatMap(exists => {
+      if (exists) {
+        if (dropIfExists) db.drop()
+        else throw new IllegalArgumentException(s"Arango Database ${db.name()} already exists")
+      }
+      db.drop()
```
[the remark this answers is not in the document]
Comment (leborgneromain) [reply persistence/src/main/scala/za/co/absa/spline/persistence/ArangoInit.scala]: fixed

File: persistence/src/main/scala/za/co/absa/spline/persistence/Persister.scala (hunk 1 above)
Comment (leborgneromain) [reply persistence/src/main/scala/za/co/absa/spline/persistence/Persister.scala]: fixed

File: integration-tests/src/main/scala/za/co/absa/spline/test/fixture/ArangoFixture.scala [hunk 5]
```diff
@@ -16,25 +16,26 @@
...
-
-  def dropAndInit(arangodb: ArangoDatabase): Unit = {
-    val db = arangodb
-    if (db.exists()) {
-      db.drop()
-    }
-    ArangoInit.initialize(db, dropIfExists = true)
+  val arangodb: ArangoDatabaseAsync = ArangoDatabaseFacade(ArangoConnectionURL(arangoUri))
+
+  def dropAndInit(arangodb: ArangoDatabaseAsync): Future[GraphEntity] = {
```
[the remark this answers is not in the document]
Comment (leborgneromain) [reply integration-tests/src/main/scala/za/co/absa/spline/test/fixture/ArangoFixture.scala]: removed that ArangoFixture from now as it is not used anyway

**Pull request merged**